File: src/chapter03_fundamentalofoop/Money.js
function Money(amount, currency) {
  if (amount < 0) {
    throw new Error('金額が0以上でありません。');
  }
  if (!currency) {
    throw new Error('通貨が正しく指定されていません。');
  }
  this.amount = amount;
  this.currency = currency;
  Object.freeze(this);  // 不変にする
}

Money.prototype.add = function(other) {
  if (this.currency !== other.currency) {
    throw new Error('通貨単位が違います。');
  }
  const added = this.amount + other.amount;
  return new Money(added, this.currency);
}

File: src/chapter03_fundamentalofoop/Money.rb
class Money
  attr_reader :amount, :currency

  def initialize(amount, currency)
    if amount < 0
      raise ArgumentError.new('金額が0以上でありません。')
    end
    if currency.nil? || currency.empty?
      raise ArgumentError.new('通貨が正しく指定されていません。')
    end
    @amount = amount
    @currency = currency
    self.freeze  # 不変にする
  end

  def add(other)
    if @currency != other.currency
      raise ArgumentError.new('通貨単位が違います。')
    end
    added = @amount + other.amount
    Money.new(added, @currency)
  end
end
